### ISSUE TITLE
Special paths in system files should be mapped as such (aka using systemd specifiers)

### ISSUE THREAD

**Issue by rugk** [k9withabone/podlet#53]
## Example

E.g. say you have the following YAML:
```yml
version: "2.1"
services:
  grocy:
    image: lscr.io/linuxserver/grocy:latest
    container_name: grocy
    volumes:
      - $HOME/config:/config:Z
    ports:
      - 9283:80
```

This currently get's mapped as (`Volume` matters here):
```ini
[Container]
ContainerName=grocy
Image=lscr.io/linuxserver/grocy:latest
PublishPort=9283:80
Volume=$HOME/config:/config:Z

[Service]
Restart=always

[Install]
WantedBy=default.target
```

## Problem
This does not work, as I dunno… some systemd thing…(?) I guess $HOME is not defined in that context.

in any case the error is something like if you want to start this Quadlet with podman:
> (podman)[82122]: grocy.service: Invalid environment variable name evaluates to an empty string: HOME/config:/config:Z

## Solution

It should replace some well-known env variables with their systemd equivalents.
`HOME` should be replaced with `%h` e.g.

Aka the result should be this:
```ini
Volume=%h/grocy-data:/config:Z
```

There are more such variables: https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#Specifiers

**Comment by k9withabone** [comment]
Podlet does not currently perform interpolation of compose files like [docker compose does](https://github.com/compose-spec/compose-spec/blob/master/spec.md#interpolation). That should probably be on the README.

I just started working on a new Rust library to better implement the compose spec and replace the `docker-compose-types` library currently used. It will take a while though. The library will eventually support interpolation.

I feel like the proposed solution is out of scope for podlet. What should also be on the README, is that podlet is mostly meant to just get you started with quadlet/systemd. I thought this was kind of implied already but I'll try to make clearer.

**Comment by k9withabone** [comment]
> So the idea is to convert all docker-compose to quadlet files and then only modify these quadlet files?

Yup, that was what I was thinking when I added the compose feature and when I created podlet originally. It doesn't really make sense to maintain a compose file if your goal is to transition to using podman and quadlets. If you want to group containers together you can use pods. Podman v5.0.0 will add support for `.pod` quadlet files. For now, if you want a configuration file to define a pod, you have to use Kubernetes YAML, which can be generated using the `podlet compose --pod` option. Note that the current functionally will be renamed to `--kube` and `--pod` will generate a `.pod` file sometime after podman v5.0.0 is released.

> I guess you're also aware of `podman-compose`

Yup. By the way, you can also use `docker compose` with podman. You just have to give it the podman Unix socket.

> Mind to explain, why, actually?

Because it feels too much like trying to guess the user's intentions. Besides, you can use a pretty simple sed command for your example: `sed 's/$HOME/%h/g'`. You may also want to try using a named volume instead of host volume so that you can just let podman worry about managing it.

**Comment by k9withabone** [comment]
Look, if you want a feature for using systemd specifiers I think you are going to have to be more specific because I'm not understanding what it is you want exactly.

- What should the option flag to activate the feature be named?
- Should it take any arguments?
- How does it change podlet's behavior?
- Which systemd specifiers should replace what values?
  - How to determine which values should be replaced?
- What kind of errors could happen?
  - How should they be handled or reported to the user?
- Any other details you feel is important for the feature.

I'd also be happy to accept a PR on this. Though how its implemented should be discussed first. It should also probably wait until after I get #52 done as implementing that will require changing the structs in the `quadlet` module. Those changes will also probably help with implementing whatever feature this turns into.

**Comment by k9withabone** [comment]
Also, just rereading your response now it sounds like you took my compose spec Rust library to mean that I'm making something like `podman-compose`. I'm not, the library is purely meant for verifying and deserializing compose files into Rust types. Maybe I shouldn't have said I'm implementing the compose spec? Though I'm not sure how else to describe it.

**Comment by rugk** [comment]
> What should the option flag to activate the feature be named?

Dunno `--use-systemd-specifiers` maybe? Or just by default?

> Should it take any arguments?

No, which ones should it?

> How does it change podlet's behavior?

Replaces unique strings (maybe start with paths only; including variables), with [the systemd specifiers defined here](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#Specifiers).

> Which systemd specifiers should replace what values? 

Yeah, maybe only paths.

> How to determine which values should be replaced?

Implementation detail. :stuck_out_tongue_winking_eye:  No just kidding, maybe expand the paths indeed and then find and replace the paths with the specifiers?

> What kind of errors could happen? 

Hmm see none? Either a replacement string is found or not. I mean replacing is an optional thing, it's not required.


---

(getting off-topic…)

> something like podman-compose

No, that's obviously clear. Though, of course you can use this project in a pipeline to convert your compose files. After all that's the aim…

> It doesn't really make sense to maintain a compose file if your goal is to transition to using podman and quadlets.

Well… depends on your use cases. `docker-compose` is a well-known format and many tools support it, so it would allow you to move away in case podman changes something or your infrastructure (requirements) change etc. https://github.com/containers/podman/discussions/20218 is full of people migrating away after `podman systemd generate` is now marked as deprecated (this [is how I came here](https://github.com/containers/podman/discussions/20218#discussioncomment-7210018)), so there is a precedent and converting Quadlet files back to `docker compose` files seems as hard as this way. I just don't want to have to migrate another time, because some tooling breaks sth.
(That all said, i think Quadlets etc. are quite cool and happily migrate, just being cautious here to keep the podman-compose.yml files.)

Anyway, enough off-topic…

**Comment by rugk** [comment]
> What does systemd do in that case?

Probably the most relevant question as this should behave the same.

It's just a quick look, but I've found these relevants spots:
* definition: https://github.com/systemd/systemd/blob/de39202426e2d3d4d6fb509635d2462b5ded6e9c/src/core/unit-printf.c#L208-L276
* replacement: https://github.com/systemd/systemd/blob/de39202426e2d3d4d6fb509635d2462b5ded6e9c/src/libsystemd/sd-path/sd-path.c#L187-L391

> Should podlet fallback to the ones from the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables)?

From what I read, yes it does, and this is the reason to use such variables in the first place.

> What if $HOME is not set?

I am not so good in C, but [apparently, it has some way more complex logic for this case than what I'd expect](https://github.com/systemd/systemd/blob/de39202426e2d3d4d6fb509635d2462b5ded6e9c/src/basic/user-util.c#L571-L606) aka e.g. having hardcoded paths for `root` and `nobody` as users.

> Specifically for the temp directories, multiple environment variables and a set directory are listed. In what order should they take precedence?

Well as I would read the doc always in reading order. [The temp dir handling as I saw in source code is more complex though](https://github.com/systemd/systemd/blob/main/src/basic/fs-util.c#L612-L651) FYI.

> What if a directory that a specifier is supposed to replace cannot be determined? I think that should be a reported error as podlet won't do what the user expects otherwise.

Yeah maybe. But given you still have a working (non-replaced/special) path, if it is not replaced, it is maybe no big deal.

> What about Windows? Should it just error if not supported?

I have no idea. Only ever used that on Linux. As such, maybe just don't offer that feature there, or as you say, error.

---

You know, this idea was born of me using this tool, just FYI. So I thought this was a good idea. If not, yeah, maybe, I just thought it would be a cool convenience feature at the time when opening the issue.

Also maybe I should clarify I would be fine with replacing _env variables_ (that have a defined mapping in systemd) for now, **not** all file paths or so. I guess (hope), this may ease implementation. I'll correct the title…

**Comment by k9withabone** [comment]
Blocked on #63 and k9withabone/compose_spec_rs#3.
